src/execution/sql/runtime_types.cpp

```diff
@@ -655,4 +658,711 @@ Timestamp Timestamp::FromYMDHMSMU(int32_t year, int32_t month, int32_t day, int3
+
+    uint128_t result_upper = half_words_result[2] | (half_words_result[3] << 64);
+    value_ = result_upper >> magic_p;
+  } else {
+    // Overflow Algorithm 2 - Magic number is > 2^128
+
+    uint128_t result_upper = half_words_result[2] | (half_words_result[3] << 64);
+    uint128_t add_upper = value_;
+
+    /*Perform addition*/
+    result_upper += add_upper;
+
+    auto carry = static_cast<uint128_t>(result_upper < add_upper);
+    carry = carry << 127;
+    // shrxi 1
+    result_upper = result_upper >> 1;
+    result_upper |= carry;
+
+    value_ = result_upper >> (magic_p - 1);
+  }
+}
+
+template <typename T>
+void Decimal<T>::UnsignedDivideConstant128Bit(uint128_t constant) {
+  if (constant == 1) {
+    return;
+  }
+
+  constexpr const uint128_t bottom_mask = (uint128_t{1} << 64) - 1;
+  constexpr const uint128_t top_mask = ~bottom_mask;
+
+  // Power of 2
+  if ((constant & (constant - 1)) == 0) {
+    uint32_t power_of_two = power_two[constant];
+    uint128_t numerator = value_;
+    numerator = numerator >> power_of_two;
+    value_ = numerator;
+    return;
+  }
+
+  // Cannot optimize if we do not have the magic number with us
+  if (magic_map128_bit_constant_division.count(constant) == 0) {
+    uint128_t numerator = value_;
+    numerator = numerator / constant;
+    value_ = numerator;
+    return;
+  }
+
+  // First input
+  uint128_t a = value_;
+
+  // Split into half words
+  uint128_t half_words_a[2];
+  uint128_t half_words_b[2];
+
+  half_words_a[0] = a & bottom_mask;
+  half_words_a[1] = (a & top_mask) >> 64;
+
+  half_words_b[0] = magic_map128_bit_constant_division[constant].lower_;
+  half_words_b[1] = magic_map128_bit_constant_division[constant].upper_;
+
+  // Calculate 256 bit result
+  uint128_t half_words_result[4];
+  // TODO(Rohan): Calculate only upper half
+  CalculateMultiWordProduct128(half_words_a, half_words_b, half_words_result, 2, 2);
+
+  uint32_t magic_p = magic_map128_bit_constant_division[constant].p_ - 128;
+
+  if (magic_map128_bit_constant_division[constant].algo_ == 0) {
+    // Overflow Algorithm 1 - Magic number is < 2^128
+
+    uint128_t result_upper = half_words_result[2] | (half_words_result[3] << 64);
+    value_ = result_upper >> magic_p;
+  } else {
+    // Overflow Algorithm 2 - Magic number is > 2^128
+
+    uint128_t result_upper = half_words_result[2] | (half_words_result[3] << 64);
+    uint128_t add_upper = value_;
+
+    /*Perform addition*/
+    result_upper += add_upper;
+
+    auto carry = static_cast<uint128_t>(result_upper < add_upper);
+    carry = carry << 127;
+    // shrxi 1
+    result_upper = result_upper >> 1;
+    result_upper |= carry;
+
+    value_ = result_upper >> (magic_p - 1);
+  }
+}
+
+template <typename T>
+void Decimal<T>::SignedMultiplyWithDecimal(Decimal<T> input, uint32_t lower_precision) {
+  bool negative_result = (value_ < 0) != (input.GetValue() < 0);
+
+  // Not used Hacker Delight 2-14 because shift needs to be agnostic of underlying T
```

Comment:
Not used??

Reply:
explaining why a particulat optimization was not done